### PR TITLE
[SPARK-36485][SQL] Support cast type constructed string as year month interval

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Spark support type constructed string as  year month interval such as 
```
interval '1 year 2 month'
interval '3 year'
```
And PGSQL support
```
cast('1 year 2 month' as interval year)
cast('1 year 2 month' as interval year to month)
cast('1 year 2 month' as interval month)
```
etc, spark can support this too.

### Why are the changes needed?
Support cast  same type constructed string as year month interval

### Does this PR introduce _any_ user-facing change?
user can cast such as `1 year 2 month` string as year month interval

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan @MaxGekk

**Review by bimanmandal — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 1]
```diff
@@ -181,8 +181,35 @@ object IntervalUtils {
               "year-month", YM(startField, endField).typeName)
           }
         }
-      case _ => throwIllegalIntervalFormatException(input, startField, endField,
-        "year-month", YM(startField, endField).typeName)
+      case _ =>
+        try {
+          val calendar = stringToInterval(input)
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:186]: Why did you implement parsing of non-ANSI interval in the function which is supposed to parse ANSI year-month interval strings?

File: sql/core/src/test/resources/sql-tests/results/cast.sql.out [hunk 2]
```diff
@@ -458,3 +458,59 @@ struct<CAST(
...
+-- !query output
+1-2
+
+
+-- !query
+SELECT CAST('1 year 2 month 3 day' AS INTERVAL YEAR TO MONTH)
+-- !query schema
+struct<CAST(1 year 2 month 3 day AS INTERVAL YEAR TO MONTH):interval year to month>
+-- !query output
+1-2
```
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/results/cast.sql.out:476]: Not sure that we should support casting of mixed interval as ANSI intervals. cc @cloud-fan

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/results/cast.sql.out:476]: @MaxGekk Do I need to continue this PR？

**Pull request closed without merging**